modules/navier_stokes/src/fvbcs/WCNSFVFluxBCBase.C

```diff
@@ -40,14 +42,80 @@ WCNSFVFluxBCBase::WCNSFVFluxBCBase(const InputParameters & params)
     _velocity_pp(isParamValid("velocity_pp") ? &getPostprocessorValue("velocity_pp") : nullptr),
     _mdot_pp(isParamValid("mdot_pp") ? &getPostprocessorValue("mdot_pp") : nullptr),
     _area_pp(isParamValid("area_pp") ? &getPostprocessorValue("area_pp") : nullptr),
-    _rho(isParamValid(NS::density) ? &getFunctor<ADReal>(NS::density) : nullptr),
+    _rho(getFunctor<ADReal>(NS::density)),
     _direction(getParam<Point>("direction")),
-    _direction_specified_by_user(params.isParamSetByUser("direction"))
+    _direction_specified_by_user(params.isParamSetByUser("direction")),
+    _vel_x(getFunctor<ADReal>(NS::velocity_x)),
+    _vel_y(isParamValid(NS::velocity_y) ? &getFunctor<ADReal>(NS::velocity_y) : nullptr),
+    _vel_z(isParamValid(NS::velocity_z) ? &getFunctor<ADReal>(NS::velocity_z) : nullptr)
 {
   if (_direction_specified_by_user && !MooseUtils::absoluteFuzzyEqual(_direction.norm(), 1.0, 1e-6))
     paramError("direction", "The direction should be a unit vector with a tolerance of 1e-6!");
 
   // Density is often set as global parameters so it is not checked
   if (_mdot_pp && _velocity_pp)
     mooseWarning("If setting the mass flow rate directly, no need for inlet velocity");
+
+  if (_subproblem.mesh().dimension() >= 2 && !_vel_y)
+    mooseError("In two or more dimensions, the y-component of the velocity must be supplied.");
+  if (_subproblem.mesh().dimension() == 3 && !_vel_z)
+    mooseError("In three dimensions, the z-component of the velocity must be supplied.");
+}
+
+// inflow is decided based on the sign of mdot or velocity postprocessors
+// depending on what is provided
+bool
+WCNSFVFluxBCBase::isInflow() const
+{
+  if (_mdot_pp)
+    return *_mdot_pp >= 0;
+  else if (_velocity_pp)
+    return *_velocity_pp >= 0;
+
+  mooseError("Either mdot_pp or velocity_pp need to be provided OR this function must be "
+             "overridden in derived classes if other input parameter combinations are valid. "
+             "Neither mdot_pp nor velocity_pp are provided.");
+  return true;
+}
+
+ADRealVectorValue
+WCNSFVFluxBCBase::varVelocity(Moose::StateArg state) const
+{
+  const auto boundary_face = singleSidedFaceArg();
+
+  ADRealVectorValue v(_vel_x(boundary_face, state));
+  if (_vel_y)
+    v(1) = (*_vel_y)(boundary_face, state);
+  if (_vel_z)
+    v(2) = (*_vel_z)(boundary_face, state);
+  return v;
+}
+
+ADReal
+WCNSFVFluxBCBase::inflowMassFlux(Moose::StateArg state) const
+{
+  if (_mdot_pp)
+    return *_mdot_pp / *_area_pp;
+  checkForInternalDirection();
```

Comment:
@grmnptr I don't think we need to check for internal boundaries at all here since it doesn't matter for determining what direction the flow goes.